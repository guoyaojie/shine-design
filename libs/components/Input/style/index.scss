/**
 * @Author 瞿龙俊 - qulongjun@shine.design
 * @Date 2018/9/17 下午7:37
 */

/**
 * 核心样式库
 */
@import "../../../sass/functions";
@import "../../../sass/config/form";
@import "../../../sass/mixins/border-radio";
@import "../../../sass/mixins/rounded-mode";
@import "../../../sass/mixins/icon";
@import "../../../sass/mixins/input";
@import "../../../sass/mixins/shadow";

/**
 * 依赖样式库
 */

/**
  独立样式
 */
@import "./input";