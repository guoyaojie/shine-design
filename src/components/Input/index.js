/**
 * @Component 按钮 - index
 * @Type 按钮组件
 * @Author 瞿龙俊 - qulongjun@shine.design
 * @Date 2019/3/15 17:57
 */

// 核心依赖
import React from 'react';

// 第三方依赖库


// 组件依赖

// 样式

export default function () {
    return (
        <input type="text" />
    );
}