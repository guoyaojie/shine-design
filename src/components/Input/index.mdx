---
name: Input 输入框
route: /guide/input
menu: form
---

import JsxParser from "react-jsx-parser";
import {
  MessageBox,
  InlineCode,
  Example,
  Code,
  PropsTable
} from "theme";
import { Input } from "shineDev";
import code from "./demo";
import "./demo/index.scss";

# Input 输入框

## 概述
在页面上生成一个表单容器，用户可以通过鼠标、键盘等设备输入内容，是最基本的表单组件。

## 代码示例

<MessageBox messageType="info">
  为了让用户直观地查看代码演示，我们加入了{" "}
  <InlineCode>shine-input-helper</InlineCode>{" "}
  修饰符，在开发过程中无需配置该字段。
</MessageBox>

#### 基础输入框

输入框组件是由基础输入标签封装而成，可以快速的创建一些常见的表单控件。

<Example 
  display={<JsxParser {...code.basic} />}
  code={{code:code.basic.jsx}}
  components={code.basic.components}
/>

#### 尺寸

输入框支持三种尺寸以适应不同的页面需求。

<Example 
  display={<JsxParser {...code.size} />}
  code={{code:code.size.jsx}}
  components={code.size.components}
/>

#### 状态

在部分应用场景下，输入框需要展示为禁用和只读状态。

<Example 
  display={<JsxParser {...code.state} />}
  code={{code:code.state.jsx}}
  components={code.state.components}
/>

#### 输入框类型

使用 `formStyle` 属性可以指定单个输入框的类型，目前支持三种类型，通过不同类型搭配可以实现个性化表单。

<Example 
  display={<JsxParser {...code.formStyle} />}
  code={{code:code.formStyle.jsx}}
  components={code.formStyle.components}
/>

#### 事件

<MessageBox messageType="info">
  打开浏览器控制台，并操作下方输入框，即可查看事件效果。
</MessageBox>

<Example 
  display={
      <Input
        placeholder="打开控制台查看效果"
        onChange={e=>{console.log('onChange',e.target.value)}}
        onBlur={e=>{console.log('onBlur',e.target.value)}}
        onClick={e=>{console.log('onClick',e.target.value)}}
      />
  }
  code={{code:code.event.jsx}}
  components={code.event.components}
/>

## 配置参数

<PropsTable of={Input} />

<MessageBox messageType="info">
  完整类型列表请参考 <a href="//developer.mozilla.org/zh-CN/docs/Web/HTML/Element/input#%E5%B1%9E%E6%80%A7">MDN</a>。
</MessageBox>
