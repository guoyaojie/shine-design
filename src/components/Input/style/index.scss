/**
 * @Author 瞿龙俊 - qulongjun@shine.design
 * @Date 2019-04-27 12:01
 */
@import "~style/mixins/shadow";
@import "~style/mixins/border-radio";
@import "~style/mixins/rounded";
@import "~style/bootstrap/input/index";

@import "./input";

@include sh-build-component--input-base();
