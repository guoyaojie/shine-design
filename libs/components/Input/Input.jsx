/**
 * @Component 输入组件 - Input
 * @Type 表单
 * @Author 瞿龙俊 - qulongjun@shine.design
 * @Date 2018/9/17 下午6:45
 */

// 核心依赖
import React, {Component, Fragment} from 'react';
import * as PropTypes from 'prop-types';

// 第三方依赖库
import classNames from 'classnames'
import _ from 'lodash';
import uuidv4 from 'uuid/v4';
// 组件依赖

// 样式
import './style';

export default class extends Component {
    constructor(props) {
        super(props);
    }

    static propTypes = {
        label: PropTypes.node,
        id: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ]),
        type: PropTypes.string,
        placeholder: PropTypes.node,
        helper: PropTypes.node,
        name: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ]),
        multiple: PropTypes.bool,
        options: PropTypes.arrayOf(PropTypes.shape({
            value: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.number
            ]),
            label: PropTypes.node,
            isChecked: PropTypes.bool,
            attributes: PropTypes.object
        })),
        attributes: PropTypes.object,
        inputStyle: PropTypes.oneOf(['pill', 'square', 'normal']),
        isAir: PropTypes.bool,
        isSolid: PropTypes.bool,
        size: PropTypes.oneOf(['sm', 'md', 'lg']),
        paddingSize: PropTypes.oneOf(['lg', 'md', 'sm']),
        display: PropTypes.oneOf(['block', 'inline']),
        labelCount: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
        isFirst: PropTypes.bool,
        isLast: PropTypes.bool,
        disabled: PropTypes.bool,
        readOnly: PropTypes.bool,
        callbacks: PropTypes.object
    };

    static defaultProps = {
        multiple: false,
        options: [],
        inputStyle: 'normal',
        isAir: false,
        isSolid: false,
        size: 'md',
        paddingSize: 'lg',
        display: 'block',
        labelCount: 3,
        isFirst: false,
        isLast: false,
        disabled: false,
        readOnly: false
    };

    render() {
        const {
            label,
            id = 'sh-' + uuidv4(),
            type,
            placeholder,
            helper,
            defaultValue,
            name,
            options,
            multiple,
            disabled,
            readOnly,
            attributes,
            inputStyle,
            isAir,
            isSolid,
            size,
            paddingSize,
            isFirst,
            isLast,
            display,
            labelCount,
            callbacks,
            className
        } = this.props;

        let _inputTyp = null;

        switch (type) {
            case 'select':
                _inputTyp = 'selectType';
                break;
            case 'textarea':
                _inputTyp = 'textareaType';
                break;
            case 'static':
                _inputTyp = 'staticType';
                break;
            default:
                _inputTyp = 'inputType'
        }

        const _className = classNames(
            'form-control',
            'sh-input',
            {
                [`sh-input--${inputStyle}`]: !_.isEqual(inputStyle, 'normal'),
                'sh-input--air': isAir,
                'sh-input--solid': isSolid,
                [`form-control-${size}`]: !_.isEqual(size, 'md'),
                [`col-${12 - labelCount}`]: _.isEqual(display, 'inline') && _.isNumber(labelCount)
            }
        );

        return (
            <Fragment>
                <div className={
                    classNames(
                        'form-group',
                        {
                            [`sh-form__group${_.isEqual(paddingSize, 'lg') ? `` : `--${paddingSize}`}`]: _.isString(paddingSize),
                            'sh-form__group--first': isFirst,
                            'sh-form__group--last': isLast,
                            'row': _.isEqual(display, 'inline')
                        },
                        ...(
                            _.isArray(className) ? className : [className]
                        )
                    )
                }>
                    {
                        _.isString(label) && (<label htmlFor={id} className={
                            classNames(
                                {
                                    'col-form-label': _.isEqual(display, 'inline'),
                                    [`col-${labelCount}`]: _.isEqual(display, 'inline') && _.isNumber(labelCount)
                                }
                            )
                        }>{label}</label>)
                    }
                    {_.isEqual(_inputTyp, 'selectType') && (
                        <Fragment>
                            <select
                                id={id}
                                className={_className}
                                name={name}
                                disabled={disabled}
                                readOnly={readOnly}
                                multiple={multiple}
                                {...attributes}
                                {...callbacks}
                            >
                                {
                                    _.isArray(options) && options.map((item) => <option value={item.value}
                                                                                        defaultChecked={item.isChecked}>{item.label}</option>)
                                }
                            </select>
                        </Fragment>
                    )}
                    {
                        _.isEqual(_inputTyp, 'textareaType') && (
                            <Fragment>
                                <textarea
                                    id={id}
                                    className={_className}
                                    name={name}
                                    disabled={disabled}
                                    readOnly={readOnly}
                                    placeholder={placeholder}
                                    defaultValue={defaultValue}
                                    {...attributes}
                                    {...callbacks}
                                />
                            </Fragment>
                        )
                    }
                    {
                        _.isEqual(_inputTyp, 'staticType') && (
                            <Fragment>
                                <p className={
                                    classNames(
                                        'form-control-static',
                                        ...(
                                            _.isArray(className) ? className : [className]
                                        ),
                                        {
                                            [`col-${12 - labelCount}`]: _.isEqual(display, 'inline') && _.isNumber(labelCount)
                                        }
                                    )
                                }>{defaultValue}</p>
                            </Fragment>
                        )
                    }
                    {
                        _.isEqual(_inputTyp, 'inputType') && (
                            <input
                                id={id}
                                type={type}
                                className={_className}
                                name={name}
                                disabled={disabled}
                                readOnly={readOnly}
                                placeholder={placeholder}
                                defaultValue={defaultValue}
                                {...attributes}
                                {...callbacks}
                            />
                        )
                    }
                    {
                        _.isString(helper) && (<span className="sh-form__help">{helper}</span>)
                    }
                </div>
            </Fragment>
        );
    }
}