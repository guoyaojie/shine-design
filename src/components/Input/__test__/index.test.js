import React from 'react';
import {render} from 'enzyme';
import Input from '../index';

describe('Input', () => {
    it('renders correctly', () => {
        const wrapper = render(<Input>Follow</Input>);
        expect(wrapper).toMatchSnapshot();
    });
});