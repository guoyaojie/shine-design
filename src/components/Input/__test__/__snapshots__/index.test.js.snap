// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "type": "text",
    },
    "children": Array [],
    "name": "input",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
